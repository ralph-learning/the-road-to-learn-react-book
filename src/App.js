import React, { Component } from 'react';
import fetch from 'isomorphic-fetch';
import PropTypes from 'prop-types';
import './App.css';

const largeColumn = { width: '40%' };
const midColumn = { width: '30%' };
const smallColumn = { width: '10%' };

const DEFAULT_QUERY = 'redux';
const DEFAULT_HITS_PER_PAGE = '10';
const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const HITS_PER_PAGE = 'hitsPerPage=';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchTerm: DEFAULT_QUERY,
      results: [],
      searchKey: '',
      error: null,
    };

    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
  }

  componentWillMount() {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(this.searchTerm);
  }

  needsToSearchTopStories(searchTerm) {
    return !this.state.results[searchTerm];
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });

    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm);
    }

    event.preventDefault();
  }

  onDismiss(id) {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotId = item => item.objectID !== id;
    const updatedHits = hits.filter(isNotId);

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page },
      }
    });
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value });
  }

  setSearchTopStories(result) {
    const { hits, page } = result;
    const { searchKey, results } = this.state;

    const oldHits = (results && results[searchKey])
      ? results[searchKey].hits
      : [];

    const updateHits = [
      ...oldHits,
      ...hits,
    ];

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updateHits, page },
      },
    });
  }

  fetchSearchTopStories(searchTerm, page = 0) {
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${HITS_PER_PAGE}${DEFAULT_HITS_PER_PAGE}`)
      .then(response => response.json())
      .then(result => this.setSearchTopStories(result))
      .catch(e => this.setState({ error: e }));
  }

  componentDidMount() {
    const  { searchTerm } = this.state;
    this.fetchSearchTopStories(searchTerm);
  }

  render() {
    const { searchTerm, results, searchKey, error } = this.state;
    const page = (
      results &&
      results[searchKey] &&
      results[searchKey].page
    ) || 0;

    const list = (
      results &&
      results[searchKey] &&
      results[searchKey].hits
    ) || [];

    return (
      <div className="page">
        <div className="interactions">
          <Search
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
            type="text"
            value={searchTerm}
          >Search</Search>
        </div>

        { error
          ? <div className="interactions">
            <p>Something went wrong.</p>
          </div>
          : <Table
            list={list}
            onDismiss={this.onDismiss}
          /> }


        <div className="interactions">
          <button onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}>
            More
          </button>
        </div>
      </div>
    );
  }
}

const Search = ({ value, onChange, onSubmit, children }) => (
  <form onSubmit={onSubmit}>
    <input
      type="text"
      value={value}
      onChange={onChange}
    />
    <button type="submit">
      {children}
    </button>
  </form>
);

Search.propTypes = {
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  OnSubmit: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
};

const Table = ({ list, onDismiss }) => (
  <div className="table">
    {list.map(item =>
      <div key={item.objectID}  className="table-row">
        <span className={largeColumn}>
          <a href={item.url}>{item.title}</a>
        </span>
        <span className={midColumn}>
          {item.author}
        </span>
        <span className={smallColumn}>
          {item.num_comments}
        </span>
        <span className={smallColumn}>
          {item.points}
        </span>
        <span className={smallColumn}>
          <Button onClick={() => onDismiss(item.objectID)}>
            Dismiss
          </Button>
        </span>
    </div> )}
  </div>
);

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number
    })
  ).isRequired,
  onDismiss: PropTypes.func.isRequired,
}

const Button = ({onClick, className, children}) => (
  <button
    onClick={onClick}
    className={className}
    type="button"
  > { children }
  </button>
);

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
}
Button.defaultProps = {
  className: '',
}

export default App;
export {
  Button,
  Search,
Table, };
