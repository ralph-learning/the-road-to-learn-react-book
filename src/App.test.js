import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App, { Search, Button, Table } from './App';

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <App />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Search', () => {
  const props = {
    OnSubmit: () => {},
    onChange: () => {},
    value: '',
  }

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Search {...props}>Search</Search>, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Search {...props}>Search</Search>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button', () => {
  const props = {
    onClick: () => {},
  }
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button {...props}>Give Me More</Button>, div);
  })

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Button {...props}>Give Me More</Button>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
})

describe('Table', () => {
  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' },
    ],
    onDismiss: () => {},
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Table {...props} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should two items in the list", () => {
    const element = shallow(
      <Table {...props} />
    );

    expect(element.find('.table-row').length).toBe(2);
  });
});


