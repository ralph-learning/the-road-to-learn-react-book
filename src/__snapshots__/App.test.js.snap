// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div
    className="table"
  >
    <div
      className="table-header"
    >
      <span
        style={
          Object {
            "width": "40%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
           
           Title 
          <i
            className="fas fa-long-arrow-alt-down"
          />
        </button>
      </span>
      <span
        style={
          Object {
            "width": "30%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
           
           Author 
          <i
            className="fas fa-long-arrow-alt-down"
          />
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
           
          Comments 
          <i
            className="fas fa-long-arrow-alt-down"
          />
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
           
           Points 
          <i
            className="fas fa-long-arrow-alt-down"
          />
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        Archive
      </span>
    </div>
  </div>
  <div
    className="interactions"
  >
    <div>
      Loading...
    </div>
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
   
  Give Me More
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <button
        className="button-inline button-active"
        onClick={[Function]}
        type="button"
      >
         
         Title 
        <i
          className="fas fa-long-arrow-alt-down"
        />
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
         
         Author 
        <i
          className="fas fa-long-arrow-alt-down"
        />
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
         
        Comments 
        <i
          className="fas fa-long-arrow-alt-down"
        />
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
         
         Points 
        <i
          className="fas fa-long-arrow-alt-down"
        />
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Archive
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      className={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href={undefined}
      >
        1
      </a>
    </span>
    <span
      className={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      className={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      className={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      className={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
         
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      className={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href={undefined}
      >
        2
      </a>
    </span>
    <span
      className={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      className={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      className={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      className={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
         
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
